File: src/magma/api/result/Result.java
package magma.api.result;

import magma.api.Tuple;

import java.util.Optional;
import java.util.function.Function;
import java.util.function.Supplier;

public interface Result<T, E extends Exception> {
    Optional<T> findValue();

    T $() throws E;

    <R> Result<R, E> mapValue(Function<T, R> mapper);

    <R> Result<R, E> flatMapValue(Function<T, Result<R, E>> mapper);

    <R> Result<Tuple<T, R>, E> and(Supplier<Result<R, E>> supplier);

    boolean isOk();
}

File: src/ts/magma/api/result/Result.ts
import { Option } from "../../../magma/api/option/Option";
export interface Result<T, X> {
	match<R>(whenOk: (arg0 : T) => R, whenErr: (arg0 : X) => R): R;
	flatMapValue<R>(mapper: (arg0 : T) => Result<R, X>): Result<R, X>;
	findError(): Option<X>;
	mapValue<R>(mapper: (arg0 : T) => R): Result<R, X>;
}
